Added class-based metrics

fsonntag opened the pull request:
With the changes in GH-75, the use of the CoNLL evaluation script was removed.
Nevertheless I found it very useful for evaluation class-based predictions, and GH-75 didn't replace this functionality.
So I added this functionality into the `Metrics` object.

Please feel free to suggest any changes and improvements.

tabergma commented:
Great idea! 
Unfortunately, your changes are breaking `Plotter.plot_training_curves()` as the tsv format changed. So we cannot merge you changes yet.
We could also reuse the adapted metric to calculate the class metrics in the text classifier. And the changes should also allow us to add the calculation of the micro and macro average score to the metric class itself. I'll take a close look in the next couple of days.

fsonntag commented:
Thanks for the feedback!
I've reverted the `to_tsv` function, as you seem to use it for plotting the process and I think it makes sense to only plot the overall metrics.
I've adjusted the `TextClassifierTrainer` and removed the `calculate_class_metrics` and the `calculate_micro_avg_metric` as they seem to be superfluous now. 
So I'm not sure if this all goes with how you have the overall architecture in mind, so please let me know if you have something else in mind!

tabergma commented on flair/training_utils.py:
```diff
@@ -15,64 +17,94 @@ class Metric(object):
+        all_classes = self.get_classes()
+        all_lines = [
+            '{0:<10}\ttp: {1} - fp: {2} - fn: {3} - tn: {4} - precision: {5:.4f} - recall: {6:.4f} - accuracy: {7:.4f} - f1-score: {8:.4f}'.format(
+                MICRO_AVG_METRIC if cls == None else cls,
```
I would take `self.name` if `cls is None` as the metric can be used in many different ways and it does not need to be the micro average of some classes.

tabergma commented on flair/training_utils.py:
```diff
@@ -15,64 +17,94 @@ class Metric(object):
+                                                     self._fns.keys()]])))
+
+        all_classes.sort(key=lambda x: (x is not None, x))
+        return all_classes
```
I would love to see methods to calculate the micro and macro average for all classes. Do you think you can add those? Otherwise I'll do it later on.

fsonntag replied:
They're added as methods to `Metric`

tabergma commented on flair/trainers/text_classification_trainer.py:
```diff
@@ -233,8 +229,7 @@ def evaluate(self, sentences: List[Sentence], eval_class_metrics: bool = False,
 
-            y_pred = []
-            y_true = []
+            metric = Metric('')
```
A metric should get a meaningful name. I would suggest either `MIRCO_AVG` or the data set type (e.g. `TEST`, `DEV` or `TRAIN`).

tabergma commented on flair/trainers/sequence_tagger_trainer.py:
```diff
@@ -215,21 +215,27 @@ def evaluate(self, evaluation: List[Sentence], out_path=None, evaluation_method:
+                        metric.fn(tag)
+                    else:
+                        metric.tn()
+                        metric.tn(tag)
```
The metrics per tag are collected but not printed at all. Do you want to add a print of the test metric after training is done?

fsonntag replied:
True, I added appropriate logging

fsonntag commented:
Thanks for the feedback! Took me a while to address the issues, but I tackled them all

tabergma commented on flair/trainers/sequence_tagger_trainer.py:
```diff
@@ -247,6 +250,16 @@ def evaluate(self, evaluation: List[Sentence], out_path=None, evaluation_method:
+            dataset_name, metric.f_score(), metric.accuracy(), metric.get_tp(), metric.get_fp(),
+            metric.get_fn(), metric.get_tn()))
+        for cls in metric.get_classes():
+            log.info("{0:<4}: f-score {1:.4f} - acc {2:.4f} - tp {3} - fp {4} - fn {5} - tn {6}".format(
```
As `metric.get_classes()` returns a list of all classes containing `None` this currently fails when training a sequence tagger due to 
```
TypeError: unsupported format string passed to NoneType.__format__
```
I would remove the `None` class from `metric.get_classes()`, so you don't log the results twice. Or you can use `dataset_name` in case of `cls is None` instead of using `cls` as name.

fsonntag replied:
I removed the `None` class from get_classes(), as it seems more consistent! Thanks!

fsonntag commented:
Did a full rebase on the master branch, everything worked fine!

tabergma commented:
Great! Thanks a lot for improving our metric class!

alanakbik commented:
@fsonntag thanks for your help!

fsonntag commented:
You're welcome, thanks for flair :)
